File: src/main/java/com/github/dkoval/leetcode/challenge/DungeonGame.java
package com.github.dkoval.leetcode.challenge;

public interface DungeonGame {

    int calculateMinimumHP(int[][] dungeon);

    // O(M * N) time | O(M * N) space
    class DungeonGameRecursiveWithMemoization implements DungeonGame {

        @Override
        public int calculateMinimumHP(int[][] dungeon) {
            int m = dungeon.length;
            int n = dungeon[0].length;
            Integer[][] memo = new Integer[m][n];
            return calculateMinimumHP(dungeon, 0, 0, memo);
        }

        private int calculateMinimumHP(int[][] dungeon, int row, int col, Integer[][] memo) {
            // idea is to solve the problem in reverse
            int m = dungeon.length;
            int n = dungeon[0].length;

            // base case: min health points needed in the bottom-right corner for the knight to survive
            if (row == m - 1 && col == n - 1) {
                return (dungeon[row][col] > 0) ? 1 : 1 - dungeon[row][col];
            }

            if (memo[row][col] != null) {
                return memo[row][col];
            }

            int answer = Integer.MAX_VALUE;

            // go right
            if (col + 1 < n) {
                int minRight = calculateMinimumHP(dungeon, row, col + 1, memo);
                answer = Math.min(answer, ensureAlive(minRight - dungeon[row][col]));
            }

            // go down
            if (row + 1 < m) {
                int minDown = calculateMinimumHP(dungeon, row + 1, col, memo);
                answer = Math.min(answer, ensureAlive(minDown - dungeon[row][col]));
            }

            memo[row][col] = answer;
            return answer;
        }

        private int ensureAlive(int healthPoints) {
            return Math.max(healthPoints, 1);
        }
    }

    // O(M * N) time | O(M * N) space
    class DungeonGameDPJava implements DungeonGame {

        @Override
        public int calculateMinimumHP(int[][] dungeon) {
            int m = dungeon.length;
            int n = dungeon[0].length;

            // hp[i][j] - min health points needed in the bottom-right for the knight to stay alive
            int[][] hp = new int[m][n];

            // base case
            hp[m - 1][n - 1] = (dungeon[m - 1][n - 1] > 0) ? 1 : 1 - dungeon[m - 1][n - 1];

            // solve for the remaining cells going bottom-up and right-to-left
            for (int i = m - 1; i >= 0; i--) {
                for (int j = n - 1; j >= 0; j--) {
                    if (i == m - 1 && j == n - 1) {
                        continue;
                    }

                    int best = Integer.MAX_VALUE;
                    // go right
                    if (j + 1 < n) {
                        best = Math.min(best, ensureAlive(hp[i][j + 1] - dungeon[i][j]));
                    }
                    // go down
                    if (i + 1 < m) {
                        best = Math.min(best, ensureAlive(hp[i + 1][j] - dungeon[i][j]));
                    }
                    hp[i][j] = best;
                }
            }
            return hp[0][0];
        }

        private int ensureAlive(int healthPoints) {
            return Math.max(healthPoints, 1);
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/DungeonGame.kt
package com.github.dkoval.leetcode.challenge

import kotlin.math.max
import kotlin.math.min

/**
 * [Dungeon Game](https://leetcode.com/explore/featured/card/june-leetcoding-challenge/541/week-3-june-15th-june-21st/3367/)
 *
 * The demons had captured the princess (P) and imprisoned her in the bottom-right corner of a dungeon.
 * The dungeon consists of M x N rooms laid out in a 2D grid.
 * Our valiant knight (K) was initially positioned in the top-left room and must fight his way through the dungeon
 * to rescue the princess.
 *
 * The knight has an initial health point represented by a positive integer. If at any point his health point
 * drops to 0 or below, he dies immediately.
 *
 * Some of the rooms are guarded by demons, so the knight loses health (negative integers) upon entering these rooms;
 * other rooms are either empty (0's) or contain magic orbs that increase the knight's health (positive integers).
 *
 * In order to reach the princess as quickly as possible, the knight decides to move only rightward or downward in each step.
 *
 * Write a function to determine the knight's minimum initial health so that he is able to rescue the princess.
 */
object DungeonGame {

    fun calculateMinimumHP(dungeon: Array<IntArray>): Int {
        val r = dungeon.size
        val c = dungeon[0].size
        val solution = Array(r) { IntArray(c) }
        // calculate min health point needed to reach the bottom-up cell
        solution[r - 1][c - 1] = if (dungeon[r - 1][c - 1] > 0) 1 else 1 - dungeon[r - 1][c - 1]
        // solve for the last column
        for (i in r - 2 downTo 0) {
            solution[i][c - 1] = ensureMinHealthPoint(solution[i + 1][c - 1] - dungeon[i][c - 1])
        }
        // solve for the last row
        for (j in c - 2 downTo 0) {
            solution[r - 1][j] = ensureMinHealthPoint(solution[r - 1][j + 1] - dungeon[r - 1][j])
        }
        // solve for remaining cells going bottom-up
        for (i in r - 2 downTo 0) {
            for (j in c - 2 downTo 0) {
                solution[i][j] = ensureMinHealthPoint(min(solution[i][j + 1], solution[i + 1][j]) - dungeon[i][j])
            }
        }
        return solution[0][0]
    }

    private fun ensureMinHealthPoint(healthPoint: Int): Int = max(healthPoint, 1)
}